### PR TITLE
chore: Store API Keys in Config File

### PR DESCRIPTION
**Pull request by onuzbee**
# Fixes #

Closes : [DI-1350](https://issues.corp.twilio.com/browse/DI-1350)
Dependent PR: [Link](https://github.com/twilio/twilio-cli/pull/259) 

Changes in profiles:create to add api keys in config file. 

Testing Gist : [Link](https://gist.github.com/onuzbee/2cfa2477ece0ce39118312428859d99c)

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [ ] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [ ] I have read the [Contribution Guidelines](https://github.com/twilio/twilio-cli-core/blob/main/CONTRIBUTING.md) and my PR follows them
- [ ] I have titled the PR appropriately
- [ ] I have updated my branch with the main branch
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation about the functionality in the appropriate .md file
- [ ] I have added inline documentation to the code I modified

If you have questions, please file a [support ticket](https://twilio.com/help/contact), or create a GitHub Issue in this repository.

### CODE REVIEW

File: src/services/config.js [hunk 1]
```diff
@@ -110,19 +120,24 @@ class ConfigData {
...
     const existing = this.getProfileById(id);
+
+    //  Update the historical projects array
     if (existing) {
       existing.accountSid = accountSid;
       existing.region = region;
     } else {
-      this.projects.push(new ConfigDataProfile(id, accountSid, region));
+      this.projects.push(new ConfigDataProject(id, accountSid, region));
     }
```
Comment (childish-sambino) [inline src/services/config.js:137]: I don't think we should be adding/updating `this.projects`. All updated profiles should be removed from `this.projects` and added to `this.profiles`. New profiles should just be going in `this.profiles`.

File: src/services/config.js [hunk 2]
```diff
@@ -119,19 +129,21 @@ class ConfigData {
...
+
+    //  Remove if existing in historical projects
     if (existing) {
-      existing.accountSid = accountSid;
-      existing.region = region;
-    } else {
-      this.projects.push(new ConfigDataProfile(id, accountSid, region));
+      this.projects = this.projects.filter((p) => p.id !== existing.id);
     }
+
```
Comment (Sindhura3) [inline src/services/config.js:144]: It could also be already existing in profiles and you might need to only update fields like region instead of creating a new profile.

File: src/services/config.js [hunk 3]
```diff
@@ -119,19 +129,21 @@ class ConfigData {
     }
   }
 
-  addProfile(id, accountSid, region) {
-    // Clean all the inputs.
+  addProfile(id, accountSid, region, apiKey, apiSecret) {
```
Comment (Sindhura3) [inline src/services/config.js:132]: Looks like addProfile was previously used only to add an entry into projects and credentials were stored in another call, but since you're doing both here, you might need to validate the key and secret inputs

File: src/services/config.js [hunk 4]
```diff
@@ -157,9 +169,8 @@ class ConfigData {
     this.prompts = configObj.prompts || {};
     // Note the historical 'projects' naming.
     configObj.projects = configObj.projects || [];
-    configObj.projects.forEach((project) => this.addProfile(project.id, project.accountSid, project.region));
-    this.setActiveProfile(configObj.activeProject);
     this.profiles = configObj.profiles || {};
```
Comment (Sindhura3) [inline src/services/config.js:172]: Along with loading, you will still need to call addProfile on each entry in the list in case there are any updates to existing ones in the file

File: src/services/config.js (hunk 3 above)
Comment (onuzbee) [reply src/services/config.js:132]: We're generating these apiKey and apiSecrets at runtime through an API call at runtime : [Link](https://github.com/twilio/twilio-cli/blob/fe17479e3b50033894d194247961083802ec403f/src/commands/profiles/create.js#L238) and shouldn't require any further validations. 

Also, just checked we seem to be saving without any further validations in keytar currently as well.

File: src/services/config.js (hunk 4 above)
Comment (onuzbee) [reply src/services/config.js:172]: We shouldn't be calling `addProfiles` on every entry on list now, since with the current change, we'll be copying everything to `profiles` on every config file read in the method. Can you please add on more here for the updates part? I might be missing something here, as In case of updates, we'll still be capturing the list changed via Line #171 ?

Also, couldn't figure out why we needed an extra step to iterate through the list when we could just load the json via `ConfigObj.projects = configObj.projects || []; `. @childish-sambino  Can you please add on more here for this step? If we need, I can add on a method `addProjects` to replicate the existing behaviour for projects.

File: src/services/config.js (hunk 2 above)
Comment (onuzbee) [reply src/services/config.js:144]: This relies on the fact that the user might still want to update the `accountSid` altogether for an existing profile as well, which'd also change the apiKey and apiSecret for it. Seemed a cleaner way to deal with it, since anyhow we'll update as per user's inputs post comparing the values already set. And, since now we're now utilising a JS Object instead of a list, we can leverage this singleton existence behaviour and not query-back and update particular fields. Thoughts? 

tldr; 
We'll be relying on user's inputs, and this doesn't create a new one, but would attempt to straight replace in case of an existing conflict.

File: src/services/config.js (hunk 4 above)
Comment (childish-sambino) [reply src/services/config.js:172]: Part of the reason for calling `this.addProfile` was to sanitize the properties. We didn't originally sanitize so some customers created profiles with leading/trailing spaces in the ID and it created problems. We updated it to sanitize when loading the profiles (and when creating profiles) to resolve this.

Also note that just storing in `configObj.projects` doesn't get the projects into `this.projects`.

File: src/services/config.js [hunk 5]
```diff
@@ -119,19 +129,21 @@ class ConfigData {
...
 
     const existing = this.getProfileById(id);
+
+    //  Remove if existing in historical projects
     if (existing) {
-      existing.accountSid = accountSid;
-      existing.region = region;
-    } else {
-      this.projects.push(new ConfigDataProfile(id, accountSid, region));
+      this.projects = this.projects.filter((p) => p.id !== existing.id);
```
Comment (ravali-rimmalapudi) [inline src/services/config.js:142]: When we are removing the existing project from `this.projects`, don't we need to remove the existing Api keys from keytar?

File: src/services/config.js [hunk 6]
```diff
@@ -103,8 +108,13 @@ class ConfigData {
       if (this.activeProfile) {
         profile = this.getProfileFromConfigFileById(this.activeProfile);
       }
+
+      // TODO: Ensure order of profiles : DI-1479
       if (!profile) {
-        profile = this.projects[0];
+        profile = this.projects[0] || Object.values(this.profiles)[0];
+        if (profile && !profile.hasOwnProperty('id')) {
```
Comment (ravali-rimmalapudi) [inline src/services/config.js:115]: `profile` will always be there because of [this](https://github.com/twilio/twilio-cli-core/pull/124/files#diff-69b5d67d841057efec105a48299840e180ffc94f0da052f105dd16f5465d6a9cR107).

File: src/services/config.js [hunk 7]
```diff
@@ -76,6 +77,10 @@ class ConfigData {
         // Clean the profile ID.
         profileId = this.sanitize(profileId);
         profile = this.getProfileFromConfigFileById(profileId);
+        // Explicitly add `id` to the returned profile
+        if (profile && !profile.hasOwnProperty('id')) {
```
Comment (ravali-rimmalapudi) [inline src/services/config.js:81]: Better to move this code snippet after Line:86,  so that we don't need [these](https://github.com/twilio/twilio-cli-core/pull/124/files#diff-69b5d67d841057efec105a48299840e180ffc94f0da052f105dd16f5465d6a9cR115) changes.

File: src/services/config.js (hunk 5 above)
Comment (onuzbee) [reply src/services/config.js:151]: Great point! 
I'll pick this up as part of [DI-1352](https://issues.corp.twilio.com/browse/DI-1352) for profile removal, and marked a comment for updation.

File: src/services/config.js (hunk 4 above)
Comment (onuzbee) [reply src/services/config.js:190]: Thanks for the details! 
Updated.

File: src/services/config.js [hunk 8]
```diff
@@ -158,8 +187,8 @@ class ConfigData {
     // Note the historical 'projects' naming.
     configObj.projects = configObj.projects || [];
     configObj.projects.forEach((project) => this.addProfile(project.id, project.accountSid, project.region));
```
Comment (childish-sambino) [inline src/services/config.js:189]: Should this call `this.addProject`?

Comment (onuzbee) [reply src/services/config.js]: Oops..Changed!

Comment (childish-sambino) [reply src/services/config.js]: Think a unit test is needed to catch this.

Comment (onuzbee) [reply src/services/config.js]: Done, added UT to assert this behaviour and keep a check on projects not being removed from list on load (undesired addProfiles behaviour). 
Rest field tests seem to be in place for Config load already.

**Review by ravali-rimmalapudi — APPROVED** [review]
Looks Good.

**Review by childish-sambino — DISMISSED** [review]

File: test/services/config.test.js [hunk 9]
```diff
@@ -248,6 +283,21 @@ describe('services', () => {
         expect(loadedConfig.getActiveProfile().id).to.equal('profile');
       });
 
+      test.it('should load projects post sanitization and not removed from list on load', async () => {
+        const config = new Config(tempConfigDir.name);
+        const configData = await config.load();
+        configData.addProfile('  profile  ', 'sid  ', '    dev', 'test_key', 'test_secret');
+        configData.addProject('    profile', ' sid ', '    dev');
```
[the remark this answers is not in the document]
Comment (onuzbee) [reply test/services/config.test.js]: Improved the assertions. 
Adding a project doesn't rely on any similar key in `profiles` and hence would be independent of it. This method is only being used to load profiles from config file to memory as is.

**Review by Sindhura3 — APPROVED** [review]
